docs/dataset.md

```diff
@@ -0,0 +1,134 @@
+# Dataset related classes
+
+## NV file format
+
+The NV file format is described in [nv-file-format](nv-file-format.md).
+
+This documentation is slightly outdated, a few fields have been added to the header.
+It remains a good introduction to understand the file layout.
+
+## Class diagram
+
+```plantuml
+@startuml
+
+package org.python.core {
+    class PySequence {}
+}
+
+package core {
+    class DatasetBase {
+        DatasetLayout layout;
+        VecBase vecMat;
+        List<DatasetRegion> regions;
+        DatasetStorageInterface dataFile;
+    }
+
+    class DatasetLayout {
+       block sizes and offsets
+    }
+    
+    class VecBase extends PySequence implements DatasetStorageInterface {}
+    
+    class DatasetRegion {}
+    
+    interface DatasetStorageInterface {}   
+    
+    class SubMatrixFile implements DatasetStorageInterface {
+        DatasetBase dataset
+        DatasetLayout layout
+    }
+    
+    class DatasetHeaderIO {
+        DatasetBase d;
+        DatasetLayout readHeader(RandomAccessFile)    
+        void writeHeader(DatasetLayout, RandomAccessFile)
+    }
+    
+    class DatasetFactory {
+        {static} newDataset(..)
+    }
+}
+
+package processor {
+    class Dataset extends DatasetBase {
+        Object analyzerObject
+    }
+    
+    class Vec extends VecBase {
+    }
+    
+    class BigMappedMatrixFile implements DatasetStorageInterface {
+        Dataset dataset
+        DatasetLayout layout
+    }
+    
+    class MappedSubMatrixFile implements DatasetStorageInterface {
+        Dataset dataset
+        DatasetLayout layout
+    }
+    
+    class MemoryFile implements DatasetStorageInterface {
+        DatasetLayout layout
+    }
+}
+
+package analyst {
+    class Analyzer
+}
+
+
+DatasetFactory ..> DatasetBase
+DatasetFactory ..> Dataset
+DatasetHeaderIO ..> DatasetBase
+DatasetBase ..> DatasetLayout
+DatasetBase ..> VecBase
+DatasetBase ..> DatasetRegion
+DatasetBase .> DatasetStorageInterface
+Dataset ..> Analyzer
+@enduml
+```
+
+## Notes
+
+### Subclassing could be simplified
+
+* Dataset is the only subclass of DatasetBase. There is no clear reason to keep them separated if we have them in the
+  same maven module.
+  * DatasetBase instances are created by:
+    * NMRStarReader. However, the instance is discarded directly after creation! Unless there is a hidden side-effect,
+      this has no use and could be removed.
+    * DatasetFactory only if Dataset is not found (should never happen when running processor or analyst)
+* Vec is the only subclass of VecBase. Same issue as between Dataset and DatasetBase
+  * VecBase constructors are only called from Vec. There is no direct usage.
```

Comment:
@brucejohnson can you confirm that those are the only usage? Or is there a weird case I missed that would create a DatasetBase or VecBase instance from python for example?